### PR TITLE
plugin: replace net/rpc plugin subsystem with gRPC implementation

### LINKED ISSUE #11: plugin: use gRPC for plugin communication
The current implementation uses the `net/rpc` interface of `go-plugin`, but we should use the improve gRPC interface

### PR DESCRIPTION
**Pull request by jrasell**
The plugin implementations are moved to gRPC, with net/rpc now a disabled protocol within the plugin manager. The plugins have been moved like-for-like, with the major differences being the use of some shared protobuf and utilities to handle the use for shared objects between different plugin interfaces.

Generation is performed by `bufbuild/buf` inline with Nomad which uses the underlying `golang/protobuf/protoc-gen-go` and has linting and sync-checks to ensure there is no drift.

The base interfaces is now more properly integrated into the plugins, removing the duplicate code that was present in the previous implementation. The base interface name has changed, bringing it inline with all the other interfaces. This will affect #353 which I will handle at the appropriate moment.

The commits have been made in an attempt to make review easier; each commit targeting a specific area for work.  

closes #11

### CODE REVIEW

**Review by lgfa29 — APPROVED** [review]
Amazing work getting this migrated! 

I left some minor comments, but they are mostly due to my lack of experience with gRPC and go-plugins.

File: plugins/shared/util_test.go [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+func Test_ActionMetaToProto(t *testing.T) {
+	testCases := []struct {
+		input               map[string]interface{}
+		expectedOutputPB    *anypb.Any
+		expectedOutputError error
+	}{
+		{
+			input: map[string]interface{}{"foo": "bar"},
+			expectedOutputPB: &anypb.Any{
+				Value: []byte{123, 34, 102, 111, 111, 34, 58, 34, 98, 97, 114, 34, 125},
```
[the remark this answers is not in the document]
Comment (jrasell) [reply plugins/shared/util_test.go:146]: I just generated them locally. Having a human readable converter func is a great idea so i'll get that added as a test helper func.

File: plugins/strategy/client.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+		Check: &sharedProto.ScalingPolicyCheck{
+			Name:        eval.Check.Name,
+			Source:      eval.Check.Source,
+			Query:       eval.Check.Query,
+			QueryWindow: ptypes.DurationProto(eval.Check.QueryWindow),
+			Strategy: &sharedProto.ScalingPolicyStrategy{
+				Name:   eval.Check.Strategy.Name,
+				Config: eval.Check.Strategy.Config,
+			},
+		},
```
[the remark this answers is not in the document]
Comment (jrasell) [reply plugins/strategy/client.go:41]: 100%...this is me forgetting/being lazy. I'll move this across into util and add some tests.

File: plugins/strategy/server.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+		Check: &sdk.ScalingPolicyCheck{
+			Name:        req.Check.GetName(),
+			Source:      req.Check.GetSource(),
+			Query:       req.Check.GetQuery(),
+			QueryWindow: queryWindow,
+			Strategy: &sdk.ScalingPolicyStrategy{
+				Name:   req.Check.Strategy.GetName(),
+				Config: req.Check.Strategy.GetConfig(),
+			},
+		},
```
[the remark this answers is not in the document]
Comment (jrasell) [reply plugins/strategy/server.go:40]: 100%...this is me forgetting/being lazy. I'll move this across into util and add some tests.

File: plugins/apm/client.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/hashicorp/nomad-autoscaler/plugins/base"
+	"github.com/hashicorp/nomad-autoscaler/plugins/shared"
+	"github.com/hashicorp/nomad-autoscaler/sdk"
+)
+
+// pluginClient is the gRPC client implementation of the APM interface.
+type pluginClient struct {
+	*base.PluginClient
+	client  proto.APMPluginServiceClient
+	doneCtx context.Context
```
[the remark this answers is not in the document]
Comment (jrasell) [reply plugins/apm/client.go:16]: It's a requirement function call param on the generated protocol buffers. I completely agree we can do a lot more with it, specifically as you point out to check whether it's an error returned by the plugin or an error making the RPC. I didn't go further into this as I wanted to limit the changes being made, so if you are happy with the idea, i'll raise a follow up to track enhancing the use of the contexts, including potential changes to our interface defs.

File: plugins/base/server.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+		pType = proto.PluginType_PLUGIN_TYPE_TARGET
+	default:
+		return nil, fmt.Errorf("plugin is of unknown type: %q", info.PluginType)
+	}
+
+	return &proto.PluginInfoResponse{
+		Type: pType,
+		Name: info.Name,
+	}, nil
+}
```
[the remark this answers is not in the document]
Comment (jrasell) [reply plugins/base/server.go:42]: TL;DR: This code translates in the opposite direction to that in the client file. I'll try and write a longer answer below; as much for me getting this out of my head as for anyone reading.

The client is responsible for calling the remote gRPC call and thus performs the following translations:
- input SDK -> input proto
- proto response -> SDK response

The server is responsible for passing the client request onto the underlying plugin implementation:
- input proto -> input SDK
- SDK response -> proto response

iirc work I have done some time ago, it is possible to forego some of the translation requirement, however, it requires the use of the generated proto code throughout the codebase which doesn't fit our requirements.

[the remark this answers is not in the document]
Comment (cgbaker) [reply plugins/base/server.go:42]: thanks for that... looking forward to reading any further notes you jot down on this

File: plugins/shared/proto/v1/shared.proto [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+message ScalingAction {
+    int64 count = 1;
+    string reason = 2;
+    bool error = 3;
+    ScalingDirection direction = 4;
+    google.protobuf.Any meta = 5;
+}
+
+enum ScalingDirection {
+    SCALING_DIRECTION_UNSPECIFIED = 0;
```
Comment (cgbaker) [inline plugins/shared/proto/v1/shared.proto:18]: what's `unspecified` for? looks like most of its uses involve an error?

**Review by cgbaker — APPROVED** [review]
great job, i appreciate the effort!

File: plugins/apm/client.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (lgfa29) [reply plugins/apm/client.go:16]: Got it, thanks for explanation. I think it makes sense to not worry about this for now. We'll can revisit this once we find the need for it.

File: plugins/base/server.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (lgfa29) [reply plugins/base/server.go:42]: Ohh...I didn't even notice they were going in opposite directions 😅 

It makes sense now, thanks for the explanation!

File: plugins/strategy/client.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lgfa29) [reply plugins/strategy/client.go:41]: Totally understandable, it's way too easy to miss things like this in PRs this size.

File: plugins/shared/proto/v1/shared.proto (hunk 6 above)
Comment (jrasell) [reply plugins/shared/proto/v1/shared.proto:18]: It is [linter rule](https://docs.buf.build/lint-rules#enum_zero_value_suffix), so it is there only to fulfil that.

**Pull request merged**